src/sftp.rs

```diff
@@ -371,11 +393,36 @@ impl Sftp {
             Err(Error::from_errno(rc))
         }
     }
+
+    fn unwrap_raw_or_err(&self) -> Result<*mut raw::LIBSSH2_SFTP, Error> {
+        match self.inner.as_ref() {
+            Some(inner) => Ok(inner.raw),
+            None => Err(Error::from_errno(raw::LIBSSH2_ERROR_BAD_USE)),
+        }
+    }
+
+    #[doc(hidden)]
+    pub fn shutdown(&mut self) -> Result<(), Error> {
+        let raw = self.unwrap_raw_or_err()?;
+        self.rc(unsafe { raw::libssh2_sftp_shutdown(raw) })?;
+        let _ = self.inner.take();
+        Ok(())
+    }
 }
 
 impl Drop for Sftp {
     fn drop(&mut self) {
-        unsafe { assert_eq!(raw::libssh2_sftp_shutdown(self.raw), 0) }
+        // Set ssh2 to blocking if sftp was not shutdown yet.
+        if let Some(inner) = self.inner.as_ref() {
```

Comment:
super minor nit, but I'd probably use `take` here rather than `as_ref` to make it extra clear that we're consuming inner